### PR TITLE
Add more tests for preprocessor #206

### PR DESCRIPTION
**Pull request by allieychen**
- Add more test cases for preprocessor.
- Changed the validation strategies. Since the contents in the report may come in any order, it is not feasible to match the exact contents. Instead, now only verifies that whether one type of error is reported by checking the headlines are present in the report. 

Issue: [issue 206](https://github.com/googlegenomics/gcp-variant-transforms/issues/206)
Tested: build and run a trigger in Container Registry.

### CODE REVIEW

File: deploy_and_run_tests.sh [hunk 1]
```diff
@@ -220,12 +220,12 @@ fi
...
-    --staging_location "gs://${gs_dir}/staging" \
-    --temp_location "gs://${gs_dir}/temp" \
-    --logging_location "gs://${gs_dir}/temp/logs" \
-    --image "${full_image_name}" ${TEST_ARGUMENTS}
+#python gcp_variant_transforms/testing/integration/run_vcf_to_bq_tests.py \
+#    --project "${project}" \
+#    --staging_location "gs://${gs_dir}/staging" \
+#    --temp_location "gs://${gs_dir}/temp" \
+#    --logging_location "gs://${gs_dir}/temp/logs" \
+#    --image "${full_image_name}" ${TEST_ARGUMENTS}
```
Comment (pgrosu) [inline deploy_and_run_tests.sh]: Why not just remove them if they're not being used?

Comment (allieychen) [reply deploy_and_run_tests.sh]: The latest commit is still in progress and it is not ready for the next round of review yet. I temporarily comment this out simply for faster testing speed. Sorry for the inconvenience. :)

Comment (pgrosu) [reply deploy_and_run_tests.sh]: No worries, totally understandable :)

File: gcp_variant_transforms/libs/preprocess_reporter.py [hunk 2]
```diff
@@ -12,7 +12,7 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-"""Generates preprocess report.
+"""Generates preprocess report (sorted).
```
Comment (arostamianfar) [inline gcp_variant_transforms/libs/preprocess_reporter.py]: nit: the 'sorted' comment is strange here as it can confuse the user reading this about what aspects are sorted. I suggest removing it or (if you really want :p ) explaining what parts are sorted.

File: gcp_variant_transforms/libs/preprocess_reporter.py [hunk 3]
```diff
@@ -175,9 +175,10 @@ def _append_malformed_records_to_report(file_to_write, malformed_records):
   """
   if malformed_records is not None:
     content_lines = []
-    for record in malformed_records:
+    sorted_malformed_records = sorted(malformed_records)
+    for record in sorted_malformed_records:
```
Comment (arostamianfar) [inline gcp_variant_transforms/libs/preprocess_reporter.py]: nit: you can just do `for record in sorted(malformed_records):`

File: gcp_variant_transforms/libs/preprocess_reporter.py [hunk 4]
```diff
@@ -233,15 +237,16 @@ def _generate_conflicting_headers_lines(
...
 
   The field ID, category (FORMAT or INFO), and the inferred header definitions
   are included in the contents.
   Output example:
   GT    INFO    num=1 type=Float
   """
   content_lines = []
-  for field_id in inferred_headers.keys():
+  sorted_inferred_header_ids = sorted(inferred_headers.keys())
+  for field_id in sorted_inferred_header_ids:
```
Comment (arostamianfar) [inline gcp_variant_transforms/libs/preprocess_reporter.py]: ditto here for inlining sorted.

File: gcp_variant_transforms/libs/preprocess_reporter.py [hunk 5]
```diff
@@ -204,24 +205,27 @@ def _generate_conflicting_headers_lines(
                     num=1 type=Integer   file2
   """
   content_lines = []
-  for field_id, definitions_to_files_map in conflicts.iteritems():
+  sorted_field_ids = sorted(conflicts.keys())
+  for field_id in sorted_field_ids:
```
Comment (arostamianfar) [inline gcp_variant_transforms/libs/preprocess_reporter.py]: in here you can also do: `for field_id in sorted(conflicts.keys()):` and the same thing below for definitions.

File: gcp_variant_transforms/testing/data/vcf/README.md [hunk 6]
```diff
@@ -34,3 +34,7 @@ FORMAT field are not defined in VCF header.
 
 `incompatible-field-value.vcf` is used to test the case where values provided
 for the fields does not match the definition. Checkout fields `NS` and `DB`.
+
+The folder `preprocessor` is created to test vcf to bq preprocessor. The
```
[the remark this answers is not in the document]
Comment (arostamianfar) [reply gcp_variant_transforms/testing/data/vcf/README.md]: Great! Thanks!

File: gcp_variant_transforms/testing/integration/run_preprocessor_tests.py [hunk 7]
```diff
@@ -76,8 +76,18 @@ def __init__(self,
             '--temp_location {}'.format(parser_args.temp_location),
             '--job_name {}'.format(
                 ''.join([test_name, suffix]).replace('_', '-'))]
+
+    self._header_blob_name = ''
+    self._header_path = ''
     for k, v in kwargs.iteritems():
-      args.append('--{} {}'.format(k, v))
+      if k == 'header_blob_name':
```
Comment (arostamianfar) [inline gcp_variant_transforms/testing/integration/run_preprocessor_tests.py]: I wonder if adding `header_blob_name=None` as part of the init arguments would work here. If not, consider adding an explicit code like this:
```
if 'header_blob_name' in kwargs:
  self._header_blob_name = self._append_suffix(kwargs['header_blob_name'])
  args.append(...)
  del kwargs['header_blob_name']
continue with the rest of the code as before.
```

File: gcp_variant_transforms/testing/integration/run_preprocessor_tests.py [hunk 8]
```diff
@@ -87,40 +97,57 @@ def validate_result(self):
     """Validates the results.
 
     - Checks that the report is generated.
-    - Validates report's contents are the same as `expected_contents`.
+    - Validates report's contents contain the strings in `expected_headlines`.
+    - Checks that the resolved headers are generated if `header_blob_name` is
+    specified in the test.
```
Comment (arostamianfar) [inline gcp_variant_transforms/testing/integration/run_preprocessor_tests.py]: nit: please indent two spaces so that `specifies` aligns with `Checks`.

File: gcp_variant_transforms/testing/integration/run_preprocessor_tests.py [hunk 9]
```diff
@@ -87,40 +97,57 @@ def validate_result(self):
...
     bucket = client.get_bucket(_BUCKET_NAME)
-    blob = bucket.get_blob(self._blob_name)
-    if not blob.exists(client):
+    report_blob = bucket.get_blob(self._report_blob_name)
+    if not report_blob:
       raise run_tests_common.TestCaseFailure(
           'Report is not generated in {}'.format(self._report_path))
-    contents = blob.download_as_string()
+    contents = report_blob.download_as_string()
     expected_contents = ''.join(self._expected_contents)
```
Comment (arostamianfar) [inline gcp_variant_transforms/testing/integration/run_preprocessor_tests.py]: I think it's easier to do `'\n'.join(self._expected_contents)` to avoid adding `\n` to every line in the test cases. WDYT?

File: gcp_variant_transforms/testing/integration/vcf_to_bq_tests/presubmit_tests/small_tests/valid_4_0.json [hunk 10]
```diff
@@ -2,7 +2,7 @@
   "test_name": "valid-4-0",
   "table_name": "valid_4_0",
   "input_pattern": "gs://gcp-variant-transforms-testfiles/small_tests/valid-4.0.vcf",
-  "runner": "DataflowRunner",
```
Comment (arostamianfar) [inline gcp_variant_transforms/testing/integration/vcf_to_bq_tests/presubmit_tests/small_tests/valid_4_0.json]: Please make this change in another PR (specific to issue #255).

File: gcp_variant_transforms/testing/integration/run_preprocessor_tests.py [hunk 11]
```diff
@@ -76,6 +77,14 @@ def __init__(self,
             '--temp_location {}'.format(parser_args.temp_location),
             '--job_name {}'.format(
                 ''.join([test_name, suffix]).replace('_', '-'))]
+    self._header_blob_name = ''
```
Comment (arostamianfar) [inline gcp_variant_transforms/testing/integration/run_preprocessor_tests.py]: nit: prefer using None instead of empty string.

File: gcp_variant_transforms/testing/integration/vcf_to_bq_tests/presubmit_tests/small_tests/valid_4_0.json (hunk 10 above)
Comment (arostamianfar) [reply gcp_variant_transforms/testing/integration/vcf_to_bq_tests/presubmit_tests/small_tests/valid_4_0.json]: Sorry, I meant leaving the small tests (unrelated to preprocess pipeline) as DataflowRunner and changing small_tests in another PR since it's unrelated to this change. You can use DirectRunner for the tests added here like you had before :)

Comment (allieychen) [reply gcp_variant_transforms/testing/integration/vcf_to_bq_tests/presubmit_tests/small_tests/valid_4_0.json]: Oh, sorry about the misunderstanding. Done now.

File: gcp_variant_transforms/testing/integration/run_preprocessor_tests.py (hunk 11 above)
Comment (allieychen) [reply gcp_variant_transforms/testing/integration/run_preprocessor_tests.py]: Done.
